dial.go

```diff
@@ -308,6 +345,84 @@ func (d FdDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
 	return conn, nil
 }
 
+// Dial makes FdDialer satisfy the Dialer interface.
+func (d FdDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	dialer := ProtocolDialer{
+		Dialer: fdDialer{
+			fd: d.Fd,
+		},
+		RequiredProtocolInfo: d.RequiredProtocolInfo,
+	}
+
+	return dialer.Dial(ctx, opts)
+}
+
+type AuthDialer struct {
+	// Dialer is a base dialer.
+	Dialer Dialer
+	// Authentication options.
+	Auth     Auth
+	Username string
+	Password string
+}
+
+// Dial makes AuthDialer satisfy the Dialer interface.
+func (d AuthDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	conn, err := d.Dialer.Dial(ctx, opts)
+	if err != nil {
+		return conn, err
+	}
+	greeting := conn.Greeting()
+	if greeting.Salt == "" {
+		conn.Close()
+		return nil, fmt.Errorf("failed to authenticate: " +
+			"an invalid connection without salt")
+	}
+
+	if d.Username == "" {
+		return conn, nil
+	}
+
+	protocolAuth := conn.ProtocolInfo().Auth
+	if d.Auth == AutoAuth {
+		if protocolAuth != AutoAuth {
+			d.Auth = protocolAuth
+		} else {
+			d.Auth = ChapSha1Auth
+		}
+	}
+
+	if err := authenticate(conn, d.Auth, d.Username, d.Password,
+		conn.Greeting().Salt); err != nil {
+		conn.Close()
+		return nil, fmt.Errorf("failed to authenticate: %w", err)
+	}
+	return conn, nil
+}
+
+type ProtocolDialer struct {
+	// Dialer is a base dialer.
+	Dialer Dialer
+	// RequiredProtocol contains minimal protocol version and
+	// list of protocol features that should be supported by
+	// Tarantool server. By default, there are no restrictions.
+	RequiredProtocolInfo ProtocolInfo
+}
+
+// Dial makes ProtocolDialer satisfy the Dialer interface.
+func (d ProtocolDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	conn, err := d.Dialer.Dial(ctx, opts)
+	if err != nil {
+		return conn, err
+	}
+	err = checkProtocolInfo(d.RequiredProtocolInfo, conn.ProtocolInfo())
```

Comment:
Here should be `identify()` call. The ProtocolDialer should return a wrapped `Conn` (protocolConn), that just proxy all methods, but returns a real `ProtocolInfo()` from `identify()`.
```suggestion
	}
	// <-
	err = checkProtocolInfo(d.RequiredProtocolInfo, conn.ProtocolInfo())
```

Reply:
Updated.